### DISCUSSION TITLE
Solver not applicable

### DISCUSSION THREAD

**Discussion by valuestay** [yalmip/YALMIP#1383] [Optimization modelling] [answered]
Dear Professor,
This is an optimization problem I encountered in reading literature. The author converted the optimization problem into a convex optimization problem and then solved it with CVX toolbox. I want to use yalmip to reproduce the code, but the error occurred, I don't know how to fix it.
A part of  literature:
![_KFLG@D5@ETUI469%WVI@XU](https://github.com/yalmip/YALMIP/assets/92216475/f97f3ccb-0fcb-4c4b-b30a-58423dfbfc8b)
my code:
[test_data.zip](https://github.com/yalmip/YALMIP/files/14645658/test_data.zip)

```
unzip('parameter.zip');
load('parameter.mat');
a_t=(abs(h_d1)+sum(abs(h_1*diag(h_r)),2)).^2;
a_r=(abs(h_d2)+sum(abs(h_2*diag(h_r)),2)).^2;

q=conj(g_1')*diag(g_0);
q_=[q g_d1'];
Q=zeros(M+1,M+1,k);
for i=1:k
    Q(:,:,i)=conj(q_(i,:)')*q_(i,:);
end

theta0_r = sdpvar(M,1,'full','complex');
t0 = sdpvar(1,1);
tk = sdpvar(k,1);
e=sdpvar(k,1);

theta0_r_=[theta0_r;1];
V0=theta0_r_*conj(theta0_r_');
W0=V0*t0;

Constraints1 = [diag(W0)==ones(size(W0,1),1)*t0,...
                           t0+sum(tk)<=T,t0>=0,tk>=0,e>=0];
for i=1:k
    Constraints1=[Constraints1,e(i)<=eta*P*trace(Q(:,:,i)*W0)];
end

Objective1=0;
for i=1:k
    Objective1=Objective1-tk(i)*log(1+e(i)*a_t(i)/tk(i)/sigma2);
end

% options = sdpsettings('solver','SDPT3');
% sol1 = optimize(Constraints1,Objective1,options); 
sol1 = optimize(Constraints1,Objective1); 

if sol1.problem == 0
    % 获取解的值
    R = -value(Objective1);
else
    disp('出错了!');
    sol1.info
    yalmiperror(sol1.problem)
end
```
error:
I thought that the optimization problem should be solved with SDPT3, but an error was reported :
```
Solver not applicable (sdpt3 does not support nonconvex quadratic terms in objective)
```
yalmip automatically assigned fmincon solver to me, and displayed 
```
fmincon stopped because the size of the current step is less than the value of the step size tolerance and constraints are 
satisfied to within the value of the constraint tolerance.
```
after iteration, and the obtained solution was incorrect.

**Comment by johanlofberg** [comment, marked as answer]
You seem to miss the core idea of a semidefinite relaxation, as you simply try to solve the nonconvex problem now, as you introduce W0 as the outer product of two vectors, i.e. a bilinear expression, which then is used in the model creating nonconvex quadratic stuff. The whole idea is that you introduce W0 as a symmetric matrix with a psd constraint, and then remove (i.e. never introduce) the constraint that it is low-rank.

SDPT3 can never be used here as you have a logarithmic operator in the objective. You need an exponential cone solver such as mosek (cvx has some old stuff where it approximates the log stuff, but these days you use an exponential cone solver)

Also, if that objective actually is exponential cone representable, you will have to rewrite it first using operators which are automatically detected as expcone representable. Just being onvex does not suffice (neither for cvx nor yalmip). From a quick glance, if this really is expcone-representable, it is likely some Kullback-Leibler term https://yalmip.github.io/command/kullbackleibler since you have terms of the for x log(1+y/x) = xlog((x+y)/x) = -xlog(x/(x+y)) = -kullbackleibler(x,x+y) or something like that

**Reply by valuestay** [reply]
Thanks for your reply. I‘m just beginning to study in convex optimization and lack knowledge and understanding in this area. The code was written step by step based on the paper, and W0 was introduced according to the description of the paper. Removing W0 and using V0 and t0 cannot obtain the correct results either. According to your guidance,I modified the objective as following：
```
Objective1=0;
for i=1:k
    Objective1=Objective1+kullbackleibler(tk(i),tk(i)+e(i)*a_t(i)/sigma2);
end
```
 Moreover, I have installed more solvers, including mosek, scs, ecos, but yalmip still chose fmincon to solve the problem, and the results are similar to before.Is there anything else that needs to be modified for constraints or objective？

**Reply by johanlofberg** [reply]
You are not supposed to remove W0 and use V0 and t0. You are missing the point with semidefinite relaxations. W0 should be defined as sdpvar(M+1) with the constraint W0 >= 0

**Reply by johanlofberg** [reply]
in the paper above, they (poorly) explain that they *replace* tau_0*V0 with a new variable called W0. Since V0 is known to be psd, a constraint W0>=0 should be added to the model but is missing in the description, and then the rank condition is removed, and a semidefinite relaxation is obtained

**Reply by valuestay** [reply]
I defined W0 and modified the objective as following:
```
W0=sdpvar(M+1,M+1,'hermitian','complex');

theta0_r_=[theta0_r;1];
V0=theta0_r_*conj(theta0_r_');

Constraints1 = [eig(W0)>=0,W0==V0*t0,diag(W0)==ones(size(W0,1),1)*t0,...
                           t0+sum(tk)<=T,t0>=0,tk>=0,e>=0];
for i=1:k
    Constraints1=[Constraints1,e(i)<=eta*P*trace(Q(:,:,i)*W0)];
end
```
However, it quickly reached the Maximum iterations. I tried to increase fmincon.MaxFunEvals, but it's not useful. As the iteration progresses, t0 always tends towards 0,which seems to mean that the iteration is not moving in the right direction. In addition, yalmip still chose to use fmincon.

**Reply by johanlofberg** [reply]
You are still missing the whole point. V0 (i.e any nonlinear product) is not part of the model (12). Relaxation means that nonlinear stuff is removed. 

And psd constraint is written as W0>=0

**Reply by valuestay** [reply]
I have tried changing `eig (W0)>=0` to `W0>=0`, and even deleted `diag (W0)==ones (size (W0,1), 1) * t0`, but the results are still incorrect.

**Reply by johanlofberg** [reply]
Why are you deleting linear stuff? 12a is kullback-leibler, 12b is linear, 12c is linear, 12d is removed as that is the whole idea in semidefinite relaxations, 12e is linear, and linear psd constraint W0>=0 is forgotten from the description but mentioned in text.

**Reply by valuestay** [reply]
I have also tried not to delete 12c, `diag (W0)==ones (size (W0,1), 1) * t0`, but the result is still incorrect.Because I couldn't get the correct results，I made many attempts，deleting 12c is one of them.😂

**Reply by johanlofberg** [reply]
And you really haven't told us what "incorrect" or "couldn't get the correct result". You are solving a semidefinite relaxation, hence there is absolutely no guarantee that the "solution" is of any relevance

**Reply by valuestay** [reply]
"Incorrect" means that in the iterated result, t0 is close to 0 and the objective value is very small, which is obviously problematic. There is another low complexity algorithm to this optimization problem, and the solution obtained by yalmip differs greatly from that obtained by this method.

**Reply by johanlofberg** [reply]
And which solver was invoked?

As I said, you are solving a relaxation, so the solution is not necessarily a solution to the original problem. YALMIP will not influence the result at all. If it is different compared to another implementation of a semidefinite relaxation, you are either solving a different relaxation, or you have implemented something incorrectly

**Reply by johanlofberg** [reply]
you are creating Q incorrectly. Per the paper, Q is q*q' (with your flipped dimensions q'*q) and should thus be hermitian and psd. Your data matrices Q(:,:,i) are not (as you counter-act the complex conjugate as you apply conj after applying ')

**Reply by valuestay** [reply]
Fmincon was invoked.
I used to think that `'` only represented transpose, so I added conj() before😂. Now I have changed Q to q' * q (my q is q^H in the paper, and the dimension of my q is 1 * (M+1)). Similarly, do I need to change `V0=theta0_r_ * conj (theta0_r_ ');`  to `V0=theta0_r_ * theta0_r_ ';`?
The current solution is much more normal, around 3.6. After the iteration was completed, it displayed：
```
Local minimum possible. Constraints satisfied.

fmincon stopped because the size of the current step is less than
the value of the step size tolerance and constraints are 
satisfied to within the value of the constraint tolerance.

<stopping criteria details>

Optimization stopped because the relative changes in all elements of x are
less than options.StepTolerance = 1.000000e-10, and the relative maximum constraint
violation, 1.009301e-10, is less than options.ConstraintTolerance = 1.000000e-06.
```
When I changed `V0=theta0'r_ * conj (theta0'r_ ');`  to `V0=theta0'r_ * theta0'r_ ', the solution became smaller, only around 1.3.

**Reply by johanlofberg** [reply]
Repeating ad nauseam *You never define or construct V0*. There is no expression or variable in the semidefinite relaxation called V0. The only decision variables are W0, tau_0, tau_k and e. 

Solving the semidefinite problem using your (fixed) code leads to mosek being invoked (i.e. it is a convex exponential cone program after relaxing) with objective -4.12

**Reply by valuestay** [reply]
OK, V0 is just an intermediate variable.But after modifing,fmincon was still invoked.

**Reply by johanlofberg** [reply]
Show the code

**Reply by valuestay** [reply]
```
a_t=(abs(h_d1)+sum(abs(h_1*diag(h_r)),2)).^2;
a_r=(abs(h_d2)+sum(abs(h_2*diag(h_r)),2)).^2;

q=conj(g_1')*diag(g_0);
q_=[q g_d1'];
Q=zeros(M+1,M+1,k);
for i=1:k
    Q(:,:,i)=q_(i,:)'*q_(i,:);
end

theta0_r = sdpvar(M,1,'full','complex');
t0 = sdpvar(1,1);
tk = sdpvar(k,1);
e=sdpvar(k,1);
W0=sdpvar(M+1,M+1,'hermitian','complex');

theta0_r_=[theta0_r;1];
V0=theta0_r_*theta0_r_';

Constraints1 = [W0>=0,W0==V0*t0,diag(W0)==ones(size(W0,1),1)*t0,...
                           t0+sum(tk)<=T,t0>=0,tk>=0,e>=0];
for i=1:k
    Constraints1=[Constraints1,e(i)<=eta*P*trace(Q(:,:,i)*W0)];
end

Objective1=0;
for i=1:k
    Objective1=Objective1+kullbackleibler(tk(i),tk(i)+e(i)*a_t(i)/sigma2);
end

sol1 = optimize(Constraints1,Objective1); 

if sol1.problem == 0
    % 获取解的值
    R = -value(Objective1);
else
    disp('出错了!');
    sol1.info
    yalmiperror(sol1.problem)
end
```

**Reply by johanlofberg** [reply]
W0==V0*t0

I cannot repeat what I have said once more...Please read my answers and the paper on what a semidefinite relaxation is

**Reply by valuestay** [reply]
I apologize for my foolishness.Now I only keep W0, tau_1, tau_k, and e as decision variables, and modify the code to
```
···
t0 = sdpvar(1,1);
tk = sdpvar(k,1);
e=sdpvar(k,1);
W0=sdpvar(M+1,M+1,'hermitian','complex');

Constraints1 = [W0>=0,diag(W0)==ones(size(W0,1),1)*t0,...
                           t0+sum(tk)<=T,t0>=0,tk>=0,e>=0];
···
```

**Reply by valuestay** [reply]
Oh,it worked,mosek was invoked successfully！！！Thank you very much for your patient answer, which helps me a lot.❤
